Position caret at end of previous block for any type of block removal

talldan opened the pull request:
## Description
As pointed out on #8805 - when removing a block (using either the keyboard shortcut or block settings menu), the caret ends up at the beginning of a preceding paragraph block. It feels natural that the cursor should instead be at the end of the block.

This ticket features a couple of simple changes to improve the situation, placing the cursor by default at the end of the the next selected block in the REMOVE_BLOCKS effect.

Some duplicated code is also removed.

There still seem to be a few cases where the flow is not quite right:
- deleting blocks using DELETE key still selects previous block (#8965)
- deleting blocks so that the first block is the selected block behaves inconsistently - sometimes the selections is the end, sometimes the start (#8963)

These issues still exist in master, so will create separate issues.

## How has this been tested?
- Test removing block(s) using block settings keyboard shortcut (event handled by BlockSettingsMenu)
- Test removing block(s) using block settings menu (event handled by BlockSettingsMenu)
- Test removing a multi-block selection using delete/backspace (event handled by EditorGlobalKeyboardShortcuts)
- Test removing a non-RichText block like the Image Block using delete/backspace (event handled by BlockListBlock

## Types of changes
<!-- What types of changes does your code introduce?  -->
Bug fix (non-breaking change which fixes an issue)
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

talldan committed:
Position caret at end of previous block for any type of block removal

talldan committed:
Remove duplicate selectBlock dispatch

- See REMOVE_BLOCKS effect in editor store's effect.js file. Final action when
removing blocks is to handle selection of the previous block

youknowriad commented:
Adding Andrew as a reviewer because he worked recently on the merge/split which also can be impacted by removal actions.

The changes look good to me, I just worry about unexpected side effects, I'll try to test a bit.

aduth commented:
The general idea here seems quite sensible on a glance to me as well. Will give a deep review shortly.

aduth commented:
Could you look in to the failing end-to-end test? I expect it may be legitimate.

Generally speaking, I'm still thinking this looks to be a very good refactor 👍

youknowriad commented:
I'm going to punt this to 3.7 :( because I don't think it's ready for 3.6

talldan commented:
e2e test seems to be passing now. I'll write a test to cover this functionality today :)

talldan committed:
Add e2e tests for various ways to delete blocks

talldan commented:
Hey @aduth & @youknowriad, I've added e2e tests, sorry for the delay.

Hope they look ok, the only way I could think to test caret position was to type some additional text to make sure it was added in the correct place. I think it should adequately guard against regressions.

aduth commented on test/e2e/support/utils.js:
```diff
@@ -272,6 +290,17 @@ export async function clickOnMoreMenuItem( buttonLabel ) {
+ */
+export async function clickOnBlockSettingsMenuItem( buttonLabel ) {
+	await expect( page ).toClick( '.editor-block-settings-menu__toggle' );
+	const itemButton = ( await page.$x( `//button[contains(text(), '${ buttonLabel }')]` ) )[ 0 ];
```
Is there any way we could improve this to limit the selector to the context of the block settings menu? Currently it could match any button on the page.

talldan replied:
Very true - I've improved the selector in a way that I think resolves this.

aduth commented on test/e2e/support/utils.js:
```diff
@@ -272,6 +290,17 @@ export async function clickOnMoreMenuItem( buttonLabel ) {
+export async function clickOnBlockSettingsMenuItem( buttonLabel ) {
+	await expect( page ).toClick( '.editor-block-settings-menu__toggle' );
+	const itemButton = ( await page.$x( `//button[contains(text(), '${ buttonLabel }')]` ) )[ 0 ];
+	await itemButton.click( 'button' );
```
Is the argument necessary? Seems `click()` would occur on the button anyways.

talldan replied:
Now removed.

aduth commented on test/e2e/support/utils.js:
```diff
@@ -272,6 +290,17 @@ export async function clickOnMoreMenuItem( buttonLabel ) {
 }
 
+/**
+ * Clicks on Block Settings Menu item, searches for the button with the text provided and clicks it.
```
Aside: Generally speaking we might consider making these utilities less specific to the specific interaction and more toward the task (triggering a block settings menu item). Opens up future refactoring, and we want to avoid expecting a particular mode of interaction (i.e. couldn't this be implemented with a keyboard? And if not, is that a problem we should be aware of?)

aduth replied:
Also, I don't think we need to rush to add functions to `utils.js`. If we're using them only in one test, it seems fine to define them in that one test, and move things as common patterns emerge.

talldan replied:
There is already a test using the keyboard shortcut for the menu item, so it might be unnecessary to also have a version that clicks the menu item.

Anyway - I've moved the code out of utils now (for this and the modifier keys mentioned above).

aduth commented on test/e2e/support/utils.js:
```diff
@@ -33,6 +33,24 @@ export const META_KEY = process.platform === 'darwin' ? 'Meta' : 'Control';
+ *
+ * @type {string}
+ */
+export const PRIMARY_ALT_MODIFIER_KEYS = process.platform === 'darwin' ? [ 'Alt', 'Meta' ] : [ 'Control', 'Alt' ];
```
Do we need this? Or is it sufficient to do `[ 'Alt', META_KEY ]` ? Same applies for the constant defined below.

talldan replied:
Removed.

aduth reviewed (commented):
Thanks for the end-to-end tests. They are very thorough! In fact, I'd expected we might only need to add a test case or two to the existing `splitting-merging.test.js` file. Upon examination of these test cases though, deletion seems to have quite a few specific behaviors, so perhaps a standalone file is appropriate.

I've left a few remarks for you to consider.

talldan committed:
Fix xpath selector for clicking on block settings menu item in e2e tests

talldan committed:
Remove unecessary param

talldan committed:
Use explicit key combination in test instead of predefining in utils

talldan committed:
Move clickOnBlockSettingsMenuItem function into the test file

talldan commented:
If #9190 is merged first, I'll have to update the new tests to use the right shortcut.

aduth commented on test/e2e/specs/block-deletion.test.js:
```diff
@@ -0,0 +1,121 @@
+			expect( await getEditedPostContent() ).toMatchSnapshot();
+		} );
+
+		it( 'positions the caret at the end of the second block as evidenced by typing additional text', async () => {
```
An `it` test should be standalone, and not make any assumptions about other `it`s being run or the order in which they're run (for example, to enable parallelization of test execution). This seems like it should be part of the previous test case.

aduth replied:
Also FWIW it's fine for a single test case to contain multiple `toMatchSnapshot`

aduth commented on test/e2e/specs/block-deletion.test.js:
```diff
@@ -0,0 +1,121 @@
+	} );
+
+	describe( 'deleting the third block using the Remove Block shortcut', () => {
+		beforeAll( addThreeParagraphsToNewPost );
```
Basically we shouldn't want to see `beforeAll( addThreeParagraphsToNewPost );` as it implies we have multiple tests operating on the same sets of blocks, which presumably are being manipulated by the individual test cases.

It should be a `beforeEach` to ensure that each test case has a consistently fresh (un-tampered) set of blocks to work with.

aduth commented on test/e2e/specs/block-deletion.test.js:
```diff
@@ -0,0 +1,121 @@
+			await page.keyboard.press( 'Backspace' );
+			expect( await getEditedPostContent() ).toMatchSnapshot();
+
+			expect( await getEditedPostContent() ).toMatchSnapshot();
```
Unnecessarily duplicated? Nothing occurs between this and the previous `toMatchSnapshot` assertion?

talldan committed:
Remove duplicate snapshot assertion

talldan committed:
Improvements to clickBlockAppender utils method

- Replace `expect( page ).toClick` with `page.click`. The former won't show
errors when used in a before/after. The latter also seems to better handle
latency issues, the former was failing when used too soon after a page load.
- Add waitForSelector to ensure the block appender is render before attempts
to click it.

talldan committed:
Ensure tests only have a single `it` statement per `describe`

talldan commented:
@aduth - thanks for the feedback. I wasn't aware test cases could be parallelized. I've combined them now. It's a shame we lose the descriptiveness of the test cases, particularly in the snapshots, but so be it.

When I made that change, the `clickBlockAppender` utils function started failing (I'm guessing some latency was introduced) so I've had to add a fix there.

talldan commented:
@aduth - If you get a moment to re-review it would be much appreciated. I believe I've addressed all the feedback.

aduth commented:
This looks great, except I'm not fond of the changes to `clickBlockAppender`. If there are latency issues, it's a bug in the application. The tests should not be updated to accommodate the buggy behavior.

We should plan to understand better the issue here. What was the error? Was the default block appender not available on the initial page load after `newPost` ? I'm not aware of what would be occurring asynchronously to be causing this.

talldan committed:
Remove waitForSelector since puppeteer does this implicitly as the first step of a call to page.click

talldan commented:
> If there are latency issues, it's a bug in the application

The issue I had was still very unusual and I wish I could explain what caused it and why it happened in the one test case and not others. The exact line that caused the latency is this one:
https://github.com/WordPress/gutenberg/pull/8961/files#diff-0f944a8b5594b301a16651bbd429262fR60

With that line present, the next test would fail. When I commented it out the next test passed. It seems very unlikely to be an issue in Gutenberg, but I am surprised it happened—the tests aren't doing anything extraordinary.

Anyway, a bit of further investigation and I've pushed a commit that removes the `waitForElement` since puppeteer does that implicitly on any `page.click` (page.click calls through to frame manager's .click):
https://github.com/GoogleChrome/puppeteer/blob/master/lib/FrameManager.js#L610

`expect.toClick` seems to use an `elementHandle` directly and that doesn't perform any implicit waiting:
https://github.com/smooth-code/jest-puppeteer/blob/master/packages/expect-puppeteer/src/matchers/toClick.js#L5

Perhaps if we don't expect any latency issues we should top using page.click.

talldan committed:
Update shortcut for remove block in block-deletion test

youknowriad commented:
This does seem ready to me, would you mind having a final look @aduth?

aduth commented:
>Anyway, a bit of further investigation and I've pushed a commit that removes the waitForElement since puppeteer does that implicitly on any page.click (page.click calls through to frame manager's .click):
https://github.com/GoogleChrome/puppeteer/blob/master/lib/FrameManager.js#L610

The changes look good. It's not immediately obvious to me by the link where Puppeteer is waiting for the element's presence? The `await` is just as a signal that it's asynchronous. The `waitForElement` would literally poll the page continuously and only resolve once the element has been found. The former assumes the element exists when it's called, the latter doesn't, despite both being asynchronous and using `await`.

Anyways, since the code looks good and tests are passing, I'm assuming it's a non-issue now.

aduth reviewed (approved):
Looks great. Thanks for the patience in the delays working through the tests for an otherwise-simple change 😄

jasmussen commented:
@talldan This is really really nice, thanks so much for working on this.

Since you've touched this, there's one other thing I would love fixed that's tangential to this. 

Right now, if you select all blocks on a page and delete them, you've lost focus. It would be nice if when deleting all blocks on the page you simply set focus on the first appender. In other words, steps to reproduce:

1. Insert a bunch of blocks. Try the demo content. 
2. Press ⌘A to select all blocks (⌘A twice if your caret is in text first)
3. Press del to delete those blocks. 

Notice how after deleting those blocks, focus is lost — it would be nice if focus was simply restored to the appender:

<img width="731" alt="screen shot 2018-08-31 at 11 34 29" src="https://user-images.githubusercontent.com/1204802/44905274-dbe3f880-ad11-11e8-9f49-86e71dca2f38.png">

aduth commented:
@jasmussen This seems like a very reasonable expectation. Do we / should we have an issue tracking it?

jasmussen commented:
@aduth I could swear I'd created one, but turns out I was thinking of #4915. Would you like me to create a new one?

talldan commented:
@jasmussen A new issue would be good 👍
